mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1332,6 +1332,120 @@ struct TosaFoldConstantMaximum
   }
 };
 
+template <typename AccumulatorType, typename InputType>
+SmallVector<AccumulatorType> matmul(
+    ShapedType outputType, ElementsAttr matrixA, ElementsAttr matrixB,
+    const std::function<AccumulatorType(const InputType &)> &convertToAccType) {
+
+  auto inputAShape = cast<ShapedType>(matrixA.getType()).getShape();
+  auto inputBShape = cast<ShapedType>(matrixB.getType()).getShape();
+
+  // InputA -> (NHC), InputB -> (NCW)
+  constexpr int64_t batchDim = 0;
+  constexpr int64_t heightDim = 1;
+  constexpr int64_t channelDim = 2;
+  constexpr int64_t widthDim = 2;
+  const auto batchSize = inputAShape[batchDim];
+  const auto channelSize = inputAShape[channelDim];
+  const auto heightSize = inputAShape[heightDim];
+  const auto widthSize = inputBShape[widthDim];
+
+  SmallVector<AccumulatorType> outputValues(outputType.getNumElements());
+  auto matrixAVals = matrixA.getValues<InputType>();
+  auto matrixBVals = matrixB.getValues<InputType>();
+
+  // Output index is always incremented by one, so avoid computing its index for
+  // each iteration.
+  auto indexOut = 0;
+  for (int64_t batch = 0; batch < batchSize; ++batch) {
+    for (int64_t height = 0; height < heightSize; ++height) {
+      for (int64_t width = 0; width < widthSize; ++width, ++indexOut) {
+        AccumulatorType acc = static_cast<AccumulatorType>(0);
+        auto indexA =
+            indexToOffset(inputAShape, {batch, height, /*channel=*/0});
+        auto indexB = indexToOffset(inputBShape, {batch, /*channel=*/0, width});
+        for (int64_t channel = 0; channel < channelSize;
+             ++channel, ++indexA, indexB += widthSize) {
+          acc += (convertToAccType(matrixAVals[indexA]) *
```

Comment:
The rules for the TOSA matmul computation seem to be rather involved 😅 Maybe you could move the addition and multiplication to some shared functionality to resemble the `apply_mul_s`/`apply_add_s` helper functions (I vote for calling them `signed` instead of just `s`)? Then they could also be used in different folder where they are required, and if they change there would only be one (easily identifiable) place which need updating?

Reply:
Hum, I was actually trying to avoid using as much function calls as possible, that is why the accumulation is done in `int64_t` and `float`, and that is also why we need the function to convert the `APInt` and `APFloat` to `int64_t`  and `float`, respectively.
I am just afraid that using function calls would slower considerably the folding.